File: node/src/tokens.ts
import { mask } from './masking.js'

const INSPECT = Symbol.for('nodejs.util.inspect.custom')

/**
 * 토큰 엔드포인트 응답을 감싼 불변 값 타입. `accessToken`/`refreshToken`은 로그·직렬화에서
 * 마스킹된다. Java/Python SDK와 동형: `idToken`(OIDC), `expiresAt`(절대 만료·epoch 초),
 * `isExpired`를 포함한다.
 */
export class TokenSet {
  constructor(
    readonly accessToken: string,
    readonly tokenType: string,
    /** 상대 만료(초). 응답의 `expires_in`. */
    readonly expiresIn: number,
    /** 절대 만료 시각(epoch **초** — JWT `exp`/Python `expires_at`과 동형). 미상이면 undefined. */
    readonly expiresAt: number | undefined,
    readonly refreshToken?: string,
    /** OIDC `id_token`(authorization_code/refresh 흐름). client-credentials에는 없다. */
    readonly idToken?: string,
    readonly scope?: string,
  ) {}

  /** `nowSeconds`/`skewSeconds`(epoch 초) 기준 만료 여부. `expiresAt` 미상이면 보수적으로 true. */
  isExpired(nowSeconds: number, skewSeconds: number): boolean {
    if (this.expiresAt === undefined) {
      return true
    }
    return nowSeconds + skewSeconds >= this.expiresAt
  }

  private masked(): string {
    return `TokenSet(tokenType=${this.tokenType}, expiresIn=${this.expiresIn}, accessToken=${mask(this.accessToken)}, refreshToken=${mask(this.refreshToken)})`
  }
  toString(): string {
    return this.masked()
  }
  toJSON(): Record<string, unknown> {
    return {
      tokenType: this.tokenType,
      expiresIn: this.expiresIn,
      expiresAt: this.expiresAt,
      accessToken: mask(this.accessToken),
      refreshToken: mask(this.refreshToken),
    }
  }
  [INSPECT](): string {
    return this.masked()
  }
}

/**
 * OIDC 토큰 응답(snake_case)을 `TokenSet`으로 매핑한다. `issuedAtSeconds`(기본 현재 시각)로부터
 * 절대 만료 `expiresAt`(epoch 초)을 계산한다 — Python `TokenSet.from_response(data, issued_at)`와 동형.
 */
export function tokenSetFromResponse(
  json: Record<string, unknown>,
  issuedAtSeconds: number = Math.floor(Date.now() / 1000),
): TokenSet {
  const at = json['access_token']
  if (typeof at !== 'string' || at.length === 0) {
    throw new Error('token response missing access_token')
  }
  const rawExpiresIn = json['expires_in']
  const expiresIn = typeof rawExpiresIn === 'number' ? rawExpiresIn : Number(rawExpiresIn ?? 0)
  const expiresAt = expiresIn > 0 ? issuedAtSeconds + expiresIn : undefined
  return new TokenSet(
    at,
    typeof json['token_type'] === 'string' ? json['token_type'] : 'Bearer',
    expiresIn,
    expiresAt,
    typeof json['refresh_token'] === 'string' ? json['refresh_token'] : undefined,
    typeof json['id_token'] === 'string' ? json['id_token'] : undefined,
    typeof json['scope'] === 'string' ? json['scope'] : undefined,
  )
}

/**
 * 강화 검증을 통과한 액세스 토큰의 신뢰 가능한 클레임. Java/Python SDK와 동형:
 * `expiresAt`(`exp`)·`issuedAt`(`iat`)을 epoch 초로 노출한다(원본 클레임은 `claims`에도 있음).
 */
export interface ValidatedToken {
  readonly subject: string
  readonly audience: readonly string[]
  readonly issuer: string
  /** `exp` 클레임(epoch 초). 없으면 undefined. */
  readonly expiresAt: number | undefined
  /** `iat` 클레임(epoch 초). 없으면 undefined. */
  readonly issuedAt: number | undefined
  readonly claims: Readonly<Record<string, unknown>>
}

/**
 * 토큰 introspection 결과. Java/Python SDK와 동형으로 `username`/`clientId`를 노출하고,
 * Node는 추가로 전체 `claims`도 노출한다(상위집합).
 */
export interface IntrospectionResult {
  readonly active: boolean
  readonly username?: string
  readonly clientId?: string
  readonly claims: Readonly<Record<string, unknown>>
}

File: dotnet/src/Xzawed.Keycloak.Sdk/Tokens.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Xzawed.Keycloak;

/// <summary>Token-endpoint response. AccessToken/RefreshToken are masked by ToString.
/// Isomorphic with the Java/Python/Node/Go TokenSet (absolute ExpiresAt in epoch seconds, IsExpired).</summary>
[JsonConverter(typeof(TokenSetJsonConverter))]   // masks access/refresh tokens in ToString() and System.Text.Json serialization.
                                                 // NOTE: reflection-based destructuring loggers (Serilog {@}) read raw
                                                 // properties and bypass this — do not @-destructure TokenSet.
public sealed record TokenSet
{
    public required string AccessToken { get; init; }
    public required string TokenType { get; init; }
    public long ExpiresIn { get; init; }        // relative seconds
    public long? ExpiresAt { get; init; }        // absolute epoch seconds; null if unknown
    public string? RefreshToken { get; init; }
    public string? IdToken { get; init; }
    public string? Scope { get; init; }

    /// <summary>Maps raw token-response values to a TokenSet, computing absolute expiry.</summary>
    public static TokenSet Create(string accessToken, string? tokenType, long expiresIn,
                                  string? refreshToken, string? idToken, string? scope, long issuedAtSeconds)
    {
        if (string.IsNullOrEmpty(accessToken))
            throw new KeycloakAuthException("token response missing access_token");
        return new TokenSet
        {
            AccessToken = accessToken,
            TokenType = string.IsNullOrEmpty(tokenType) ? "Bearer" : tokenType,
            ExpiresIn = expiresIn,
            ExpiresAt = expiresIn > 0 ? issuedAtSeconds + expiresIn : null,
            RefreshToken = refreshToken,
            IdToken = idToken,
            Scope = scope,
        };
    }

    /// <summary>Conservative: an unknown ExpiresAt is treated as expired.</summary>
    public bool IsExpired(long nowSeconds, long skewSeconds) =>
        ExpiresAt is null || nowSeconds + skewSeconds >= ExpiresAt.Value;

    public override string ToString() =>
        $"TokenSet {{ TokenType = {TokenType}, ExpiresIn = {ExpiresIn}, " +
        $"AccessToken = {Masking.Mask(AccessToken)}, RefreshToken = {Masking.Mask(RefreshToken)} }}";
}

/// <summary>Trusted claim set of an access token that passed hardened validation.</summary>
public sealed record ValidatedToken(
    string Subject,
    IReadOnlyList<string> Audience,
    string Issuer,
    long? ExpiresAt,
    long? IssuedAt,
    IReadOnlyDictionary<string, object?> Claims);

/// <summary>RFC 7662 introspection response.</summary>
public sealed record IntrospectionResult(
    bool Active,
    string? Username,
    string? ClientId,
    IReadOnlyDictionary<string, object?> Claims);

/// <summary>Returned by CreateAuthorizationRequest to start a PKCE authorization-code flow.
/// The caller stores CodeVerifier/State/Nonce until the callback (the SDK is stateless).
/// CodeVerifier is masked by ToString and by System.Text.Json — it is the proof-of-possession
/// secret for the code exchange, so a stolen code plus a logged verifier completes the flow.
/// ⚠️ Being a positional record, the compiler-generated ToString would otherwise print every
/// property; the override below is what prevents that. Url/State/Nonce stay visible, matching
/// Rust's Debug impl (only code_verifier is masked there too).</summary>
[JsonConverter(typeof(AuthorizationRequestJsonConverter))]
public sealed record AuthorizationRequest(string Url, string CodeVerifier, string State, string Nonce)
{
    public override string ToString() =>
        $"AuthorizationRequest {{ Url = {Url}, State = {State}, Nonce = {Nonce}, " +
        $"CodeVerifier = {Masking.Mask(CodeVerifier)} }}";
}

/// <summary>Masks the PKCE verifier when an AuthorizationRequest is JSON-serialized.
/// NOTE: same caveat as TokenSet — reflection-based destructuring loggers (Serilog {@}) read raw
/// properties and bypass this converter; do not @-destructure AuthorizationRequest.</summary>
internal sealed class AuthorizationRequestJsonConverter : JsonConverter<AuthorizationRequest>
{
    public override AuthorizationRequest Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        => throw new NotSupportedException("AuthorizationRequest is not deserializable from JSON.");

    public override void Write(Utf8JsonWriter writer, AuthorizationRequest value, JsonSerializerOptions options)
    {
        writer.WriteStartObject();
        writer.WriteString("url", value.Url);
        writer.WriteString("state", value.State);
        writer.WriteString("nonce", value.Nonce);
        writer.WriteString("codeVerifier", Masking.Mask(value.CodeVerifier));
        writer.WriteEndObject();
    }
}

/// <summary>Masks access/refresh tokens when a TokenSet is JSON-serialized via System.Text.Json.
/// NOTE: reflection-based destructuring loggers (Serilog {@}) read raw properties and bypass this
/// converter entirely — do not @-destructure TokenSet.</summary>
internal sealed class TokenSetJsonConverter : JsonConverter<TokenSet>
{
    public override TokenSet Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        => throw new NotSupportedException("TokenSet is not deserializable from JSON.");

    public override void Write(Utf8JsonWriter writer, TokenSet value, JsonSerializerOptions options)
    {
        writer.WriteStartObject();
        writer.WriteString("tokenType", value.TokenType);
        writer.WriteNumber("expiresIn", value.ExpiresIn);
        if (value.ExpiresAt is { } ea) writer.WriteNumber("expiresAt", ea); else writer.WriteNull("expiresAt");
        writer.WriteString("accessToken", Masking.Mask(value.AccessToken));
        writer.WriteString("refreshToken", Masking.Mask(value.RefreshToken));
        writer.WriteEndObject();
    }
}
